### PR TITLE
[quant][core][feature] Implement masked_fill for CUDA tensors

### PR DESCRIPTION
**Pull request by jcaip**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #85164
* __->__ #85108

Summary:
- Add new cuda test for masked_fill
- Add in QuantizedCUDA implementation for masked_fill

Test Plan:
```
python test/test_quantization.py -k test_qtensor_masked_fill_cuda

```

Reviewers: dzdang

Subscribers:

Tasks: Fixes https://github.com/pytorch/pytorch/issues/83110

Tags: quant

### CODE REVIEW

File: test/quantization/core/test_quantized_tensor.py [hunk 1]
```diff
@@ -1032,8 +1037,7 @@ def test_qtensor_masked_fill(self):
                 [numel], scale=scale, zero_point=zero_point, device=device,
                 dtype=qtype)
             # mask fill the whole tensor, equivalent to calling plain vanilla fill
-            mask = torch.tensor(True)
-            torch.tensor(fill_with)
+            mask = torch.tensor(True, device=device)
```
Comment (vadimkantorov) [inline test/quantization/core/test_quantized_tensor.py:1040]: does masked_fill_ also work with python scalar? e.g. q_masked_fill.masked_fill_(True, fill_with)

Comment (jcaip) [reply test/quantization/core/test_quantized_tensor.py:1040]: I don't think so, `masked_fill` expects a tensor mask, so it will throw a type error.

Comment (vadimkantorov) [reply test/quantization/core/test_quantized_tensor.py:1040]: my bad, i got it mistaken for the fill value

File: test/quantization/core/test_quantized_tensor.py [hunk 2]
```diff
@@ -1049,18 +1053,19 @@ def test_qtensor_masked_fill(self):
...
         for qtype, fill_with in itertools.product(types, fills):
             q_masked_fill = qx.clone()
             q_masked_fill.masked_fill_(mask, fill_with)
             ref = qx.clone()
+
             for i in range(numel):
                 if mask[i]:
                     # this assignment doesn't end up calling masked_fill, allowing us to compare the different implementations
-                    ref[i] = fill_with
+                    ref[i] = torch.tensor([fill_with], device=device, dtype=torch.float)
```
Comment (dzdang) [inline test/quantization/core/test_quantized_tensor.py:1068]: btw you can also just `do torch.tensor(fill_with, device=device, dtype=torch.float)`

**Review by dzdang — APPROVED** [review]
lgtm

File: test/quantization/core/test_quantized_tensor.py [hunk 3]
```diff
@@ -1049,18 +1053,19 @@ def test_qtensor_masked_fill(self):
 
         # the above loop does the same test as test_qtensor_fill
         # now we will check masked_fill for subset of indices
-        mask = torch.randint(0, 2, (numel, ))
+        mask = torch.randint(0, 2, (numel, ), device=device)
```
Comment (vadimkantorov) [inline test/quantization/core/test_quantized_tensor.py:1056]: related : https://github.com/pytorch/pytorch/issues/35527
